fix: allow_to_create_new_org checks org autocreate [FC-0076]

pomegranited opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Updates the StudioHome API's allow_to_create_new_org to require _both_ organization-creation permissions and `settings.ORGANIZATIONS_AUTOCREATE` to be enabled.

Previously, users with staff permission were allowed to _attempt_ organization autocreation when creating a course or library using the Authoring MFE, but if `ORGANIZATIONS_AUTOCREATE` was disabled, they would get an error from the creation API.

[`ORGANIZATIONS_AUTOCREATE` is enabled by default](https://github.com/openedx/edx-platform/blob/c5fabd586376478b8c5b784c3bd7ce5264c259dd/cms/envs/common.py#L2767), but edx.org disables it.

Useful information to include:

- Which edX user roles will this change impact? Course Authors

## Supporting information

Part of: https://github.com/openedx/frontend-app-authoring/issues/1577

## Testing instructions

See Authoring MFE PR.

## Deadline

None

pomegranited committed:
fix: allow_to_create_new_org checks org autocreate

Updates the StudioHome API's allow_to_create_new_org to require both
organization-creation permissions and ORGANIZATION_AUTOCREATE to be
enabled.

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!


This repository is currently maintained by @openedx/wg-maintenance-edx-platform.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary><h4>:radio_button: Get product approval</h4></summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details><details>
<summary><h4>:radio_button: Provide context</h4></summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details><details>




<summary><h4>:radio_button: Get a green build</h4></summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<summary><h3>Where can I find more information? </h3></summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details><details>
<summary><h3>When can I expect my changes to be merged?</h3></summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited commented on cms/djangoapps/contentstore/rest_api/v1/views/home.py:
```diff
@@ -80,7 +81,12 @@ def get(self, request: Request):
 
         home_context = get_home_context(request, True)
         home_context.update({
-            'allow_to_create_new_org': settings.FEATURES.get('ENABLE_CREATOR_GROUP', True) and request.user.is_staff,
+            # 'allow_to_create_new_org' is actually about auto-creating organizations
+            # (e.g. when creating a course or library), so we add an additional test.
+            'allow_to_create_new_org': (
+                home_context['can_create_organizations'] and
+                org_api.is_autocreate_enabled()
+            ),
```
For reference, the `can_create_organizations` value is set here:

https://github.com/openedx/edx-platform/blob/c5fabd586376478b8c5b784c3bd7ce5264c259dd/cms/djangoapps/contentstore/views/course.py#L1794-L1798

which is pretty different from the original `allow_to_create_new_org`. But it didn't make sense to me that we'd use a different default value for `settings.FEATURES['ENABLE_CREATOR_GROUP']` AND require the user to be staff here.

navinkarkera replied:
@pomegranited This seems reasonable but the logic while creating courses seems different. If the setting to create new org is set to `False`, the API only allows user/staff to create courses in organizations they are allowed in: 
* https://github.com/open-craft/frontend-app-course-authoring/blob/a25a09e053ea0e0d153bfd28280072ebe6e5475f/src/generic/create-or-rerun-course/hooks.jsx#L35
* https://github.com/open-craft/edx-platform/blob/1ac6e02bb8a16ac76bc0f2e4d25626e4e0734c01/cms/djangoapps/contentstore/utils.py#L1582

We also have [get_allowed_organizations_for_libraries](https://github.com/open-craft/edx-platform/blob/1ac6e02bb8a16ac76bc0f2e4d25626e4e0734c01/cms/djangoapps/contentstore/utils.py#L1581) for libraries, so probably we need to do the same here.

pomegranited replied:
Good catch @navinkarkera !  see [f1f40e0](https://github.com/openedx/edx-platform/pull/36094/commits/f1f40e0e87fdd37aa147770fa3b197175832c50f)

pomegranited committed:
fix: enforce a user's "allowed organizations" when creating libraries

Also adds the list of "allowed organizations for libraries" to the
Studio Home API, so it can be used by the Authoring MFE.

navinkarkera reviewed (approved):
:+1: Looks good! @pomegranited 
- [x] I tested this: (Tried creating org via new lib page and course page)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

ChrisChV committed:
Merge branch 'master' into jill/studio-home-api-enforce-org-autocreate

hinakhadim commented:
I'm from 2u-TnL team. Recently, we noticed that we are not able to create Course (being staff users) because we are not getting organizations in CreateCourseForm. In my devstack, I'm able to do so and everything is working fine. I've gone through this PR and found that `settings.ORGANIZATIONS_AUTOCREATE` should be True. In devstack locally, it is already true. You have also mentioned in the description that "[ORGANIZATIONS_AUTOCREATE is enabled by default](https://github.com/openedx/edx-platform/blob/c5fabd586376478b8c5b784c3bd7ce5264c259dd/cms/envs/common.py#L2767), but edx.org disables it." 

Is there any reason that this flag is False for edx.org?

pomegranited commented:
Hi @hinakhadim , my apologies for a delayed reply and for the disruption this change has caused your team.

> You have also mentioned in the description that "[ORGANIZATIONS_AUTOCREATE is enabled by default](https://github.com/openedx/edx-platform/blob/c5fabd586376478b8c5b784c3bd7ce5264c259dd/cms/envs/common.py#L2767), but edx.org disables it."

Yep -- I got this information from @kdmccormick on the original issue: https://github.com/openedx/frontend-app-authoring/issues/1577 (and there's more background about it in [this ADR](https://github.com/openedx/edx-organizations/blob/master/docs/decisions/0001-phase-in-db-backed-organizations-to-all.rst)). But given that context, I think it's important to keep `ORGANIZATIONS_AUTOCREATE == False` on edx.org.

> we are not able to create Course (being staff users) because we are not getting organizations in CreateCourseForm

Ok, this seems to be the real issue -- you _should_ see the list of organizations that you're allowed to create courses for in that list. But from digging in today, I think this PR conspired with [this line in the Course Authoring MFE](https://github.com/openedx/frontend-app-authoring/blob/8fe52d22e70586f6049d7368a22bd790adec45fe/src/generic/create-or-rerun-course/hooks.jsx#L35) -- so instead of seeing the full list of organizations, now you only see the "allowedOrganizations" list. But that [allowed_organizations](https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/views/course.py#L1749-L1756) list is empty because `FEATURES["ENABLE_CREATOR_GROUP"] == False`. Unfortunately I can't recommend that you set [`FEATURES["ENABLE_CREATOR_GROUP"] = True`](https://github.com/openedx/edx-platform/blob/a2bb8c94586ab15a28c5f6b360c8504d432bcbc4/cms/envs/common.py#L222-L224) either -- that flag would change a lot of things for a lot of people.

So I think we need to fix [the Course Authoring MFE](https://github.com/openedx/frontend-app-authoring/blob/8fe52d22e70586f6049d7368a22bd790adec45fe/src/generic/create-or-rerun-course/hooks.jsx#L35) like this: https://github.com/openedx/frontend-app-authoring/pull/1689

hinakhadim commented:
@pomegranited Thank you for your response and for identifying the issue correctly.

For our stage instance, `FEATURES["ENABLE_CREATOR_GROUP"]` is set to True, while `ORGANIZATIONS_AUTOCREATE` remains False.

Since `ENABLE_CREATOR_GROUP` is enabled, the function is executing as expected. However, as I am a Staff user rather than a Course Creator, I am receiving an empty array ([]) from [here](https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/views/course.py#L1794-L1796) in the code. 

Furthermore, I am sharing a screenshot of the **Stage Studio API** of my user for further clarity. You can see `allow_to_create_org` is false and `can_create_organizations` is true. I believe your fix PR will address these conditions and resolve the issue.

<img width="787" alt="image" src="https://github.com/user-attachments/assets/fc7248b0-d91c-48f9-b46e-f6eb03c1fcd6" />
